R/MsBackendCached.R

```diff
@@ -0,0 +1,616 @@
+        "Number of rows in local data and number of spectra don't match"
+    else NULL
+}
+
+#' @rdname MsBackendCached
+#'
+#' @export
+MsBackendCached <- function() {
+    new("MsBackendCached")
+}
+
+setValidity("MsBackendCached", function(object) {
+    msg <- c(.valid_local_data(object@localData, object@nspectra))
+    ## For local data that are core spectra variables, check for valid data type
+    msg <- c(msg, .valid_column_datatype(object@localData))
+    if (is.null(msg)) TRUE
+    else msg
+})
+
+#' @rdname MsBackendCached
+setMethod("backendInitialize", "MsBackendCached",
+          function(object, data = data.frame(), nspectra = 0L,
+                   spectraVariables = character(), ...) {
+              if (nrow(data) == 0L)
+                  data <- data.frame(matrix(ncol = 0, nrow = nspectra))
+              else nspectra <- nrow(data)
+              object@nspectra <- as.integer(nspectra)
+              object@localData <- data
+              object@spectraVariables <- spectraVariables
+              validObject(object)
+              object
+})
+
+#' @rdname MsBackendCached
+setMethod("dataStorage", "MsBackendCached", function(object) {
+    rep("<cache>", length(object))
+})
+
+#' @rdname MsBackendCached
+setMethod("length", "MsBackendCached", function(x) {
+    x@nspectra
+})
+
+#' @rdname MsBackendCached
+setMethod("spectraVariables", "MsBackendCached", function(object) {
+    unique(c(names(.SPECTRA_DATA_COLUMNS), colnames(object@localData),
+             object@spectraVariables))
+})
+
+#' @importFrom S4Vectors make_zero_col_DFrame
+#'
+#' @noRd
+.spectra_data <- function(x, columns = spectraVariables(x)) {
+    local_cols <- intersect(columns, colnames(x@localData))
+    ## @spectraVariables are supposed to be added by the extending backend
+    columns <- columns[!columns %in% x@spectraVariables]
+    core_cols <- intersect(columns, names(.SPECTRA_DATA_COLUMNS))
+    core_cols <- core_cols[!core_cols %in% c(local_cols, x@spectraVariables)]
+    res <- NULL
+    if (length(local_cols))
+        res <- as(x@localData[, local_cols, drop = FALSE], "DataFrame")
+    if (length(core_cols)) {
+        mzvals <- NULL
+        intvals <- NULL
+        lst <- NumericList(numeric(), compress = FALSE)
+        if (any(core_cols == "mz")) {
+            core_cols <- core_cols[core_cols != "mz"]
+            mzvals <- lst[rep(1, times = length(x))]
+        }
+        if (any(core_cols == "intensity")) {
+            core_cols <- core_cols[core_cols != "intensity"]
+            intvals <- lst[rep(1, times = length(x))]
+        }
+        if (length(core_cols))
+            tmp <- DataFrame(lapply(.SPECTRA_DATA_COLUMNS[core_cols],
+                                    function(z, n) rep(as(NA, z), n), length(x)))
+        else tmp <- make_zero_col_DFrame(x@nspectra)
+        tmp$mz <- mzvals
+        tmp$intensity <- intvals
+        if (length(res))
+            res <- cbind(res, tmp)
+        else res <- tmp
```

Comment:
I don't really get this part. Why adding mz/intensity to `res` when it is always `NULL` or a `NumericList` with an `integer(0)` per row. Is this useful at all?

Reply:
So you mean why having the `tmp$mz <- mzvals` line when `mzvals` is `NULL`?

The reasoning of that code is:
- if there is any column `"mz"` (or `"intensity"`) requested, then initialize the variabl `mzvals` and **remove** the requested column.
- create a `DataFrame` with the requested columns and `NA` values of the correct data type.
- add `$mz` to this data frame. If it was not requested `NULL` (nothing) will be added. Otherwise the `NumericList` is added. This is to avoid having yet another `if` statements for mz and intensity.
 
This whole code just takes care to add *default* (empty) values for core spectra variables.

Reply:
Exactly.  This function is a convenience function for extending backends that now don't have to implement the code to return empty (`NA`) core spectra variables.